src/hypothesis/extra/django/models.py

```diff
@@ -17,159 +17,359 @@
+add_default_field_mapping(dm.EmailField, email_field_values)
+
+
+float_field_values = _simple_field_strategy(
+    st.floats,
+    allow_nan=False,
+    allow_infinity=False,
+)
+add_default_field_mapping(dm.FloatField, float_field_values)
+
+
+integer_field_values = _simple_field_strategy(
+    st.integers,
+    min_value=-2147483648,
+    max_value=2147483647,
+)
+add_default_field_mapping(dm.IntegerField, integer_field_values)
+
+
+null_boolean_field_values = _simple_field_strategy(
+    partial(st.one_of, st.none(), st.booleans()),
+)
+add_default_field_mapping(dm.NullBooleanField, null_boolean_field_values)
+
+
+positive_integer_field_values = _simple_field_strategy(
+    st.integers,
+    min_value=0,
+    max_value=2147483647,
+)
+add_default_field_mapping(
+    dm.PositiveIntegerField,
+    positive_integer_field_values,
+)
+
+
+positive_small_integer_field_values = _simple_field_strategy(
+    st.integers,
+    min_value=0,
+    max_value=32767,
+)
+add_default_field_mapping(
+    dm.PositiveSmallIntegerField,
+    positive_small_integer_field_values,
+)
+
+
+slug_field_values = _simple_field_strategy(
+    model_text,
+    alphabet=string.digits + string.ascii_letters + '-',
+)
+add_default_field_mapping(dm.SlugField, slug_field_values)
+
+
+small_integer_field_values = _simple_field_strategy(
+    st.integers,
+    min_value=-32768,
+    max_value=32767,
+)
+add_default_field_mapping(dm.SmallIntegerField, small_integer_field_values)
+
+
+time_field_values = _simple_field_strategy(
+    times,
+    timezones=(),
+)
+add_default_field_mapping(dm.TimeField, time_field_values)
+
+
+url_field_values = _fake_factory_field_strategy(u'uri')
+add_default_field_mapping(dm.URLField, url_field_values)
+
+
+def _resolve_strategy_factory(strategy, *args, **kwargs):
+    if callable(strategy):
+        return strategy(*args, **kwargs)
     return strategy
 
 
-def models(model, **extra):
-    result = {}
-    mandatory = set()
-    for f in model._meta.concrete_fields:
-        try:
-            strategy = _get_strategy_for_field(f)
-        except UnmappedFieldError:
-            mandatory.add(f.name)
+@defines_strategy
+def field_values(field, **kwargs):
+    """A strategy of valid values for the given field."""
+    try:
+        strategy = _field_mappings[type(field)]
+    except KeyError:
+        strategy = _simple_field_strategy(st.nothing)
+    strategy = _resolve_strategy_factory(strategy, field, **kwargs)
+    return strategy
+
+
+# Model strategies.
+
+@defines_strategy
+def models(model, __db=st.none(), **field_strategies):
+    """A strategy of valid models, saved to the database."""
+    # Allow strategy factories.
+    field_strategies = {
+        field_name: _resolve_strategy_factory(
+            strategy,
+            model._meta.get_field(field_name),
+        )
+        for field_name, strategy
+        in field_strategies.items()
+    }
+    # Introspect model for extra fields.
+    for field in model._meta.concrete_fields:
```

Comment:
This should be changed to a .get_fields() call with a filter for f.is_concrete, concrete_fields is an [internal API](https://github.com/django/django/blob/master/django/db/models/options.py#L501).